### PR TITLE
Handle connection timeout

### PR DESCRIPTION
**Pull request by sandipbhoir**
## Description
This PR addresses issue - https://github.com/Zilliqa/Issues/issues/439

## Backward Compatibility

- [x] This is not a breaking change
- [ ] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [x] small-scale cloud test ( 33 node )
- [x] large-scale cloud test ( 1200 nodes )

### CODE REVIEW

**Comment by kaikawaliu** [comment]
How about using epoll in our case, I feel it may bring better performance for our use case. Ref: https://devarea.com/linux-io-multiplexing-select-vs-poll-vs-epoll/#.XNj-z3VuZhE

**Comment by kaikawaliu** [comment]
and also https://jvns.ca/blog/2017/06/03/async-io-on-linux--select--poll--and-epoll/

**Comment by sandipbhoir** [comment]
epoll may not fit in this context ( for solving connection timeout of individual client connection )
epoll will be useful when we create multiple client connection before hand and likely to wait on it until fd are ready or timeout.

So i think poll fits here well because with current design we open new connection when we have something to be sent to specific client.

Unless, we redesign complete networking to have open connection to all peers in the list which has its own challenge because sending port has to be dynamic since we need multiple connection between same sender and receiver in parallel.

**Comment by bb111189** [comment]
While this will prevent incoming connection from fully consume all FD and unable to get new FD, I think another angle we want to look at leaves some buffer for FD, else other parts of Zilliqa may not work as intended (eg. reading of JSON from smart contract, writing to log file). 

Here is what I found for checking the FD limit . 

> For the current process count, you can use getrlimit to get the file descriptor limit, then iterate over all integers from 0 to that limit and try calling fcntl with the F_GETFD command. It will succeed only on the file descriptors which are actually open, letting you count them.


https://stackoverflow.com/questions/7976769/getting-count-of-current-used-file-descriptors-from-c-code

**Comment by sandipbhoir** [comment]
Bit confused. I don't see this PR preventing any fd's used by incoming connection. It is only for outgoing connection.

**Comment by bb111189** [comment]
> While this will prevent incoming connection from fully consume all FD and unable to get new FD, I think another angle we want to look at leaves some buffer for FD, else other parts of Zilliqa may not work as intended (eg. reading of JSON from smart contract, writing to log file).
> 
> Here is what I found for checking the FD limit .
> 
> > For the current process count, you can use getrlimit to get the file descriptor limit, then iterate over all integers from 0 to that limit and try calling fcntl with the F_GETFD command. It will succeed only on the file descriptors which are actually open, letting you count them.
> 
> https://stackoverflow.com/questions/7976769/getting-count-of-current-used-file-descriptors-from-c-code

Resolved. sorry for the confusion

File: src/libNetwork/P2PComm.cpp [hunk 1]
```diff
@@ -204,22 +209,93 @@ bool SendJob::SendMessageSocketCore(const Peer& peer, const bytes& message,
...
+    }
+    arg |= O_NONBLOCK;
+    if (fcntl(cli_sock, F_SETFL, arg) < 0) {
+      LOG_GENERAL(WARNING, "couldn't set socket as non-blocking");
+      return false;
+    }
+
+    int status;
+    bool connectStat = true;
+    if ((status = connect(cli_sock, (struct sockaddr*)&serv_addr,
```
Comment (ansnunez) [inline src/libNetwork/P2PComm.cpp]: It looks like if `connect()` fails then `connectStat` remains `true`.  Is the flow for that case OK?

**Pull request merged**